accounts-db/src/accounts_db.rs

```diff
@@ -3230,19 +3263,27 @@ impl AccountsDb {
         let useful_accum = AtomicU64::new(0);
 
         // parallel scan the index.
-        let (mut purges_zero_lamports, purges_old_accounts) = {
+        let purges_old_accounts = {
             let do_clean_scan = || {
                 candidates
-                    .par_chunks(4096)
-                    .map(|candidates: &[Pubkey]| {
-                        let mut purges_zero_lamports = HashMap::new();
+                    .par_iter()
+                    .map(|candidates_bin| {
                         let mut purges_old_accounts = Vec::new();
                         let mut found_not_zero = 0;
                         let mut not_found_on_fork = 0;
                         let mut missing = 0;
                         let mut useful = 0;
+                        // Must separate keys for passing as an argument to accounts_index.scan(),
+                        // otherwise two simultaneous, conflicting borrows occur
+                        // 1. immutable borrow of bin to iterate over the keys, and
+                        // 2. mutable borrow of bin to update bin values in the callback closure.
+                        // An alternative is to wrap the bin values in RefCell for non-exclusive
+                        // access.
+                        let candidate_keys: Vec<Pubkey> =
+                            candidates_bin.read().unwrap().keys().copied().collect();
```

Comment:
one idea is:
```
let candidates = candidates_bin.read().unwrap();
candidates.iter_mut().for_each(|k, v| {
index.scan([pk].iter(), || same closure using mutable v inside,
...)
});
```
this avoids the copy. it means scan is called on individual entries. it probably doesn't do anything fancy anyway.

Reply:
`index.scan([pk].iter() ...)` means that we make a call to `scan` for every single `pubkey`, correct?

Reply:
Do we need to use HashMap to store the candidates? It seems that we are only iterating on the candidates. 
If that's the case, can we use two vec instead, - one for pubkeys, one for CleaningInfo?
Then, for scan, we can pass pubkeys vec as immutable ref, and CleaningInfo as mutable ref to callback?

Reply:
Yes, I think this should work. Let me try this.

Reply:
The problem with separate keys and cleaning info values is that AccountsIndex::scan passes to the callback a pubkey, not the index of the pubkey in a vector of pubkeys. It's possible to change this, but I don't know in how many other places scan is used and the callbacks expect to map a pubkey to something. This may be a pervasive change.

Reply:
yeah. And I think another options is to maintain the index outside of `scan` and in the callback, increase index. 
```
                        let mut index = 0;
                        self.accounts_index.scan(
                            pubkeys.iter(),
                            |pubkey, slots_refs, _entry| {
                                let info = &mut clean_infos[index];
                                index += 1;
                                ...
                            },
                            None,
                            false,
                        );
```

Reply:
`index += 1;` We do this in other places. scan does iterate in order.

Reply:
> `index.scan([pk].iter() ...)` means that we make a call to `scan` for every single `pubkey`, correct?

yes. This means we call `scan` once per key instead of calling `scan` once with an iter of pubkeys. But, it also means we don't have to make a copy of the pubkeys first. I think the index lookup is way worse than the call to `scan`. We could also make a version of `scan` which only works on a single pubkey, but I imagine the overhead of just calling `scan` once per pubkey is irrelevant. Avoiding the pubkey copies is nice. But, it also only happens once per bin in parallel and can then be dropped, so maybe it isn't that big of a deal. easy enough to measure the cost of creating the pubkey vecs. just add up all the time in a metric, run ledger tool, dump it out so we have some idea of how much time it takes. we already know how much memory it takes.

Reply:
I'm reworking this to separate pubkeys and cleaning info vectors.

Reply:
It seems just too much trouble to split the HashMap into separate vectors of pubkeys and cleaning info, because it is actually used as a hashmap in calc_delete_dependencies and filter_zero_lamport_clean_for_incremental_snapshots, and it also makes the code more convoluted.  I think it's better go with Jeff's suggestion of calling scan once per key.

Reply:
Yes. Agree. scan one key at a time sounds good to me too.